### ISSUE TITLE
smoothing of amplitudes leaves high outliers

### ISSUE THREAD

**Issue by rvweeren** [lofar-astron/factor#31]
Wondering why these amplitude large outliers, see example, are not taken out by the amplitude smoothing code.

<img width="153" alt="rs210outliers" src="https://cloud.githubusercontent.com/assets/3879473/14057930/b4a1623c-f2e5-11e5-9907-26d07010378e.png">

**Comment by AHorneffer** [comment]
Smoothing is currently done by a couple of 1-d median-filters only along the time axis.
- Should we do 2-d smoothing on the time-frequency plane?
- Should we do smoothing by fitting a function to the data? (E.g. a 3-rd oder polynomial for each night?)
- Are there other suggestions?

**Comment by rvweeren** [comment]
I noted that the cut is at 6 sigma. Maybe that is simply too high and we need to lower it ?
(I can try to experiment a bit this week….)
- Including 2D smoothing also sounds like a good idea, we might catch more bad stuff that way (although often the bad time ranges are the same for different blocks). Also it requires at at least ~6 neighboring blocks to get statistics along the freq. axis, so maybe we should only trigger it if there are a sufficient number of neighboring freq. blocks.
- Fitting might also be worthwhile, maybe there are better functions than polynomials. An oscillatory  function that in the limit of t --> +/-infinity goes to 1.0 would be ideal.

**Comment by rvweeren** [comment]
I had a look at changing the cut from 6sigma. I concluded we should not lower it (it creates  problems when the number of solution time lots is low and it starts over smoothing).

**Comment by darafferty** [comment]
I see that LoSoTo uses [scipy.interpolate.UnivariateSpline](http://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.UnivariateSpline.html) for smoothing clock solutions. I wonder it that could work for our amplitudes?

**Comment by rvweeren** [comment]
Yes, that scipy.interpolate.UnivariateSpline looks very useful. At least to determine outliers from that best fitting model which we can then replace by the Spline fit.

**Comment by rvweeren** [comment]
A 2D spline (if we have enough points along the freq. axis) might be even better. That should add additional safeguards against bad data.

[scipy 2D spline](http://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.BivariateSpline.html#scipy.interpolate.BivariateSpline)

**Comment by rvweeren** [comment]
Have been experimenting with LSQUnivariateSpline. Attached figure shows the spline fits, works quite well. I still have to optimize the number of knots based on the amplitude scatter (i.e., if there is more scatter less control points should be used, so the fit is forced to be more smooth).

![test](https://cloud.githubusercontent.com/assets/3879473/14358428/e3a21b06-fcba-11e5-92c3-d3b007d6c03b.png)

(also trying 2D splines, but still fighting with array data types)

**Comment by rvweeren** [comment]
Now with automatic detection of outliers.

![test](https://cloud.githubusercontent.com/assets/3879473/14362093/0abbeeb8-fccc-11e5-87d8-2a7ba74b7d42.png)

**Comment by rvweeren** [comment]
And testing it on some really bad data (where essential there no useful solution on the distant RS stations). The identification of bad data points seems pretty stable to me.

(I coded it such that the spline order, i.e. smoothness, scales with the intrinsic scatter of the data points and number of data points available). So it fits more smooth splines when the scatter is large. (fitting is done in log-space and iterates twice)

![test](https://cloud.githubusercontent.com/assets/3879473/14362614/505615dc-fcce-11e5-875d-0296005dd7f7.png)

**Comment by rvweeren** [comment]
Added a simple form of "adaptive" knot control to deal with locally varying noise. (noise is indicated by the dashed lines, normalized to a mean of 1, i.e, the scatter gives an absolute measure).

Attached some examples

Apart from some more testing and code clean up this is as good as it gets with 1D splines I think.

![test](https://cloud.githubusercontent.com/assets/3879473/14390530/8c8df65c-fd85-11e5-8fd5-483836e9e44a.png)

![test](https://cloud.githubusercontent.com/assets/3879473/14390588/d4be36d0-fd85-11e5-88c1-1d6666d2244f.png)

![test](https://cloud.githubusercontent.com/assets/3879473/14390643/18935a2a-fd86-11e5-9432-40792bff396d.png)

**Comment by darafferty** [comment]
This looks like a really nice improvement over the current median filter.  It should reduce the number of free parameters a bit, too, though probably not significantly.

**Comment by rvweeren** [comment]
Cleaned up the 1D spline code, working on the 2D one now (that's a bit more complicated).

At the moment the code just detects the outliers and replaces them with the spline fit. (the current smoothcal code also replaces only the outliers points). If we really trust the spline fits we could think about replacing the full solutions with the actual spline fit. That would reduce the number of "free amplitude parameters" by about about an order of magnitude or so.

**Comment by rvweeren** [comment]
Question: Does anyone know how to get the freq axis values out of a parmdb ?

**Comment by rvweeren** [comment]
Ah already found the answer:

parms[key_names[0]]['freqs']

**Comment by rvweeren** [comment]
Can anyone provide a _blabla.merge_amp_parmdbs2_ out of one of a _facet_patch_nr_ directory made with at least 15+ blocks (preferable even more, i.e., the full bandwidth) ?

(I need that for 2D spline fitting tests)

**Comment by twshimwell** [comment]
Sure, I have a few with the full bandwidth. I will put them in https://www.dropbox.com/s/mhir6g43zk7l9pm/for_reinout.tar?dl=0

**Comment by rvweeren** [comment]
Thanks, some first simple 2D spline fits (no outlier rejection yet or any more complicated stuff, for example bad data points still affect the fit now and I've not yet included data weights). I have to determine what trends (along the freq axis) are real and which ones are caused by bad data...

(I am more used to looking at amplitude time series, so will have to develop some feeling for what changes are expected along the freq. axis)

![test](https://cloud.githubusercontent.com/assets/3879473/14505975/774b3fc2-0189-11e6-83b3-65ae96bc7a1f.png)

**Comment by AHorneffer** [comment]
On 2016-04-13 21:17, Reinout van Weeren wrote:

> (I am more used to looking at amplitude time series, so will have to develop
> some feeling for what changes are expected along the freq. axis)

I would put time on the x-axis and frequency on the y-axis in the plots. I 
believe that will be easier to read.

   Andreas

**Comment by rvweeren** [comment]
I am used to this (AIPS, casapy) …. but rfigui does it indeed the other way (how you prefer it).

Note that this is for a "short" time axis. (freq-axis is always constant but the time-axis varies). For high S/N sources the time axis becomes very long (1000 pixels for 3C295). For that reason I put time on the y-axis. The question is do you prefer a very wide figure (we need 3 images with 1000 pixels on the x-axis then) or a very long one? I think most people prefer to scroll up/down on a figure, not left/right, and have a variable figure hight (not a variable width).

But, it will be easy to change it. This was just a first test plot to verify the code was doing something sensible.

**Comment by rvweeren** [comment]
1D fits to the parmdbs Tim provided (had to tweak/fix a few things, related to flagged stations)

(I am just showing the solutions for 1 "freq channel" of the parmdb)

The reason why it finds many outlier for the low-amp parts of CS030HBA1 (first plot) is that the amplitudes get close to zero, and hence in log-space the scatter becomes very large. It then replaces the points with the fit. This is the desired behavior (because correcting with low-amp values with blow up the data when doing a correct step, so this should stabilize the correction for low-gain data if you want to image). The first plot must be for an outlier source (or at the edge of the nominal FoV) given the large variations for all stations (sources going in/out of the primary beam lobe).

![test](https://cloud.githubusercontent.com/assets/3879473/14548259/2e4ea9a8-0283-11e6-8bff-a6c7c48d59b9.png)
![test](https://cloud.githubusercontent.com/assets/3879473/14548339/03ba87c4-0284-11e6-8071-0e80ad4b752c.png)
![test](https://cloud.githubusercontent.com/assets/3879473/14548491/354f69ac-0285-11e6-9654-37690166956e.png)

**Comment by rvweeren** [comment]
After a few days trying abandoned the 2D spline, the changes on the freq axis are not well captured by it. Adopted a 1D-spline fit along the time axis, followed by a 2D median filter model.

![test2d](https://cloud.githubusercontent.com/assets/3879473/14550387/455f14ee-0296-11e6-84b7-e4a04b887da8.png)

**Comment by rvweeren** [comment]
And for one of Tim's parmdb (source at the edge of the FoV), with more freq. channels.

![test2d](https://cloud.githubusercontent.com/assets/3879473/14550521/4df60ab2-0297-11e6-91c4-44815ca0e7ac.png)

**Comment by rvweeren** [comment]
And a final one for today, again one from Tim (but now with more "normal" solution). 

![test2d](https://cloud.githubusercontent.com/assets/3879473/14550561/acfa9eb0-0297-11e6-8357-3f0b20fcf6d2.png)

**Comment by darafferty** [comment]
These results look really good. What about the slow phases? Could we use something similar to smooth them?

**Comment by rvweeren** [comment]
In principle yes, BUT, the problem is that you have to trust the TEC solutions. (because if they are wrong the slow phases will try to compensate them). And that is what I see happening. Image shows the slow phases for RS210. (this was for 3C295 so I had enough S/N to check the "slow" phase corrections on the same timescale as the TEC to check if the are wrong TEC solutions. If you increase the "slow" solution interval to longer values often these issues often become hidden and they do not show the clear step functions.  So the jumps are -not- wrong slow phase solutions, but it is trying to correct errors in the TEC solutions. So we have to fix our TEC issues first before we can deal with the slow phases.

(I've seen quite many examples (from others) with "bad" slow phases, in many cases that is not because of poor S/N, but again a reflection of their bad TEC solutions. In fact, for quite a number of directions I notice high-frequency ripple patterns that are not calibrated out, again this must be due to the TEC convergence problems. I think this is the last remaining major -fundamental- problem that has to be addressed)

<img width="184" alt="rs210" src="https://cloud.githubusercontent.com/assets/3879473/14565938/6a556392-02fb-11e6-97d9-b336d71c619e.png">

**Comment by darafferty** [comment]
I've added an option to the reimage branch to enable the spline + 2D median smoothing. Set  "spline_smooth2D = True" under [calibration] to enable it.

**Comment by rvweeren** [comment]
As far as I can tell the new spline and 2D amplitude outlier rejection works. I'll close this.
